source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,208 @@
+#include "common/filter/http/filter_config_discovery_impl.h"
+
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/utility.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Filter {
+namespace Http {
+
+DynamicFilterConfigProviderImpl::DynamicFilterConfigProviderImpl(
+    FilterConfigSubscriptionSharedPtr&& subscription,
+    const std::set<std::string>& require_type_urls,
+    Server::Configuration::FactoryContext& factory_context)
+    : subscription_(std::move(subscription)), require_type_urls_(require_type_urls),
+      tls_(factory_context.threadLocal().allocateSlot()),
+      init_target_("DynamicFilterConfigProviderImpl", [this]() {
+        subscription_->start();
+        init_target_.ready();
```

Comment:
I've spent 5-10 minutes looking at this and I'm pretty confused by the various init targets at play here. Naively this seems wrong, shouldn't we become ready when the first subscription success/failure is received? I'm probably missing something. Can you add more comments around the init target usages in this file and in the header file?